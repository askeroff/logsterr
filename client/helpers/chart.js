const checkHasData = projects => {
  let hasData = false;
  projects.forEach(project => {
    if (project.seconds) {
      hasData = true;
    }
  });
  return hasData;
};

export const getDrilldown = data => {
  const result = [];

  const recursive = (children, name) => {
    const childrenResult = [];
    let parentId;
    children.forEach(child => {
      const hasData = checkHasData(child.children) && child._id;
      if (child.seconds) {
        parentId = child.parent_id;
        childrenResult.push({
          name: child.name,
          y: child.seconds / 3600,
          drilldown: hasData || null
        });
      }
      if (hasData) {
        recursive(child.children, child.name);
      }
    });
    if (childrenResult.length) {
      result.push({
        name,
        id: parentId,
        drilldown: null,
        data: childrenResult
      });
    }
  };

  data.forEach(project => {
    if (project.seconds && checkHasData(project.children)) {
      recursive(project.children, project.name);
    }
  });
  return result;
};

export const prepareData = (data, drilldown) => {
  const seriesData = data
    .map(item => {
      if (item.seconds && checkHasData(item.children)) {
        return {
          name: item.name,
          y: item.seconds / 3600,
          drilldown: item._id
        };
      } else if (item.seconds) {
        return {
          name: item.name,
          y: item.seconds / 3600,
          drilldown: null
        };
      }
      return null;
    })
    .filter(item => item !== null);
  return {
    xAxis: {
      type: 'category'
    },
    series: [
      {
        name: 'Projects',
        colorByPoint: true,
        data: seriesData
      }
    ],
    drilldown: {
      series: drilldown
    }
  };
};
