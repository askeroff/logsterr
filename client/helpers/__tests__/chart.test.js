import { getDrilldown, prepareData } from '../chart';
import data from './mockData';

test('Drilldown is the right shape', () => {
  const drilldown = getDrilldown(data);
  expect(drilldown).toMatchSnapshot();
});

test('Prepare data is correct', () => {
  const drilldown = getDrilldown(data);
  const prepared = prepareData(data, drilldown);
  expect(prepared).toMatchSnapshot();
});

