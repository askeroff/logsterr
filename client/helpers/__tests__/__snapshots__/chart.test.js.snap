// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drilldown is the right shape 1`] = `
Array [
  Object {
    "data": Array [
      Object {
        "drilldown": null,
        "name": "Project 2",
        "y": 16.546944444444446,
      },
      Object {
        "drilldown": null,
        "name": "Project 5",
        "y": 0.36666666666666664,
      },
    ],
    "drilldown": null,
    "id": "5a281df108567500ade59253",
    "name": "Project 1",
  },
  Object {
    "data": Array [
      Object {
        "drilldown": null,
        "name": "Project 10",
        "y": 3.143611111111111,
      },
      Object {
        "drilldown": null,
        "name": "Learning Songs",
        "y": 4.273055555555556,
      },
    ],
    "drilldown": null,
    "id": "5a2a1703a2efd310b8225595",
    "name": "Project 8",
  },
]
`;

exports[`Prepare data is correct 1`] = `
Object {
  "drilldown": Object {
    "series": Array [
      Object {
        "data": Array [
          Object {
            "drilldown": null,
            "name": "Project 2",
            "y": 16.546944444444446,
          },
          Object {
            "drilldown": null,
            "name": "Project 5",
            "y": 0.36666666666666664,
          },
        ],
        "drilldown": null,
        "id": "5a281df108567500ade59253",
        "name": "Project 1",
      },
      Object {
        "data": Array [
          Object {
            "drilldown": null,
            "name": "Project 10",
            "y": 3.143611111111111,
          },
          Object {
            "drilldown": null,
            "name": "Learning Songs",
            "y": 4.273055555555556,
          },
        ],
        "drilldown": null,
        "id": "5a2a1703a2efd310b8225595",
        "name": "Project 8",
      },
    ],
  },
  "series": Array [
    Object {
      "colorByPoint": true,
      "data": Array [
        Object {
          "drilldown": "5a281df108567500ade59253",
          "name": "Project 1",
          "y": 24.85638888888889,
        },
        Object {
          "drilldown": null,
          "name": "Project 7",
          "y": 1.2355555555555555,
        },
        Object {
          "drilldown": "5a2a1703a2efd310b8225595",
          "name": "Project 8",
          "y": 7.416666666666667,
        },
        Object {
          "drilldown": null,
          "name": "Project 11",
          "y": 2.799722222222222,
        },
      ],
      "name": "Projects",
    },
  ],
  "xAxis": Object {
    "type": "category",
  },
}
`;
